File: app/services/PacketToJsonTransfer.java
package services;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import org.pcap4j.packet.*;
import play.libs.Json;

import javax.inject.Singleton;
import java.sql.Timestamp;

@Singleton
public class PacketToJsonTransfer {

    public JsonNode packageToJson(Packet packet, Timestamp timestamp) {
        ObjectNode jsonNode = Json.newObject();
        jsonNode.put("timestamp", timestamp.toString());

        if (packet.contains(EthernetPacket.class)) {
            jsonNode.set("ethernet",
                    getEthernetPacketMetrics(packet.get(EthernetPacket.class)));
        }
        if (packet.contains(ArpPacket.class)) {
            jsonNode.set("arp",
                    getArpPacketMetrics(packet.get(ArpPacket.class)));
        }
        if (packet.contains(IpPacket.class)) {
            jsonNode.set("ip",
                    getIpPacketMetrics(packet.get(IpPacket.class)));
        }
        if (packet.contains(TcpPacket.class)) {
            jsonNode.set("tcp",
                    getTcpPacketMetrics(packet.get(TcpPacket.class)));
        }
        if (packet.contains(UdpPacket.class)) {
            jsonNode.set("udp",
                    getUdpPacketMetrics(packet.get(UdpPacket.class)));
        }
        return jsonNode;
    }

    private JsonNode getEthernetPacketMetrics(
            EthernetPacket ethernetPacket) {
        ObjectNode jsonNode = Json.newObject();
        EthernetPacket.EthernetHeader header = ethernetPacket.getHeader();
        jsonNode.put("macSrcAddr", header.getSrcAddr().toString());
        jsonNode.put("macDstAddr", header.getDstAddr().toString());
        jsonNode.put("etherType", header.getType().name());
        return jsonNode;
    }

    private JsonNode getArpPacketMetrics(ArpPacket arpPacket) {
        ObjectNode jsonNode = Json.newObject();
        ArpPacket.ArpHeader header = arpPacket.getHeader();
        jsonNode.put("srcHardwareAddr", header.getSrcHardwareAddr().toString());
        jsonNode.put("srcProtocolAddr", header.getSrcProtocolAddr().toString());
        jsonNode.put("dstHardwareAddr", header.getDstHardwareAddr().toString());
        jsonNode.put("dstProtocolAddr", header.getDstProtocolAddr().toString());
        jsonNode.put("hardwareType", header.getHardwareType().name());
        jsonNode.put("operation", header.getOperation().name());
        jsonNode.put("protocolType", header.getProtocolType().name());
        return jsonNode;
    }

    private JsonNode getIpPacketMetrics(IpPacket ipPacket) {
        ObjectNode jsonNode = Json.newObject();
        IpPacket.IpHeader header = ipPacket.getHeader();
        jsonNode.put("srcAddr", header.getSrcAddr().toString());
        jsonNode.put("dstAddr", header.getDstAddr().toString());
        jsonNode.put("protocol", header.getProtocol().name());
        jsonNode.put("version", header.getVersion().name());
        return jsonNode;
    }

    private JsonNode getTcpPacketMetrics(TcpPacket tcpPacket) {
        ObjectNode jsonNode = Json.newObject();
        jsonNode.put("srcPort", "" + tcpPacket.getHeader().getSrcPort().valueAsInt());
        jsonNode.put("srcPortName", "" + tcpPacket.getHeader().getSrcPort().name());
        jsonNode.put("dstPort", "" + tcpPacket.getHeader().getDstPort().valueAsInt());
        jsonNode.put("dstPortName", "" + tcpPacket.getHeader().getDstPort().name());
        return jsonNode;
    }

    private JsonNode getUdpPacketMetrics(UdpPacket udpPacket) {
        ObjectNode jsonNode = Json.newObject();
        jsonNode.put("srcPort", "" + udpPacket.getHeader().getSrcPort().valueAsInt());
        jsonNode.put("srcPortName", "" + udpPacket.getHeader().getSrcPort().name());
        jsonNode.put("dstPort", "" + udpPacket.getHeader().getDstPort().valueAsInt());
        jsonNode.put("dstPortName", "" + udpPacket.getHeader().getDstPort().name());
        return jsonNode;
    }

}

File: app/services/PacketToJsonTransfer.scala
package services

import java.sql.Timestamp

import org.pcap4j.packet._
import play.api.libs.json.{JsString, JsValue, Json}

/**
  * Utility class that extracts data from network packets and puts them into JSON
  *
  * Created by Kristian Lange on 2017.
  */
object PacketToJsonTransfer {

  def packageToJson(packet: Packet, timestamp: Timestamp): JsValue = {
    var json = Json.obj()
    json += ("timestamp", JsString(timestamp.toString))
    if (packet.contains(classOf[EthernetPacket]))
      json += ("ethernet", getEthernetPacketMetrics(packet.get(classOf[EthernetPacket])))
    if (packet.contains(classOf[ArpPacket]))
      json += ("arp", getArpPacketMetrics(packet.get(classOf[ArpPacket])))
    if (packet.contains(classOf[IpPacket]))
      json += ("ip", getIpPacketMetrics(packet.get(classOf[IpPacket])))
    if (packet.contains(classOf[TcpPacket]))
      json += ("tcp", getTcpPacketMetrics(packet.get(classOf[TcpPacket])))
    if (packet.contains(classOf[UdpPacket]))
      json += ("udp", getUdpPacketMetrics(packet.get(classOf[UdpPacket])))
    json
  }

  private def getEthernetPacketMetrics(ethernetPacket: EthernetPacket) = Json.obj(
    "macSrcAddr" -> ethernetPacket.getHeader.getSrcAddr.toString,
    "macDstAddr" -> ethernetPacket.getHeader.getDstAddr.toString,
    "etherType" -> ethernetPacket.getHeader.getType.name
  )

  private def getArpPacketMetrics(arpPacket: ArpPacket) = Json.obj(
    "srcHardwareAddr" -> arpPacket.getHeader.getSrcHardwareAddr.toString,
    "srcProtocolAddr" -> arpPacket.getHeader.getSrcProtocolAddr.toString,
    "dstHardwareAddr" -> arpPacket.getHeader.getDstHardwareAddr.toString,
    "dstProtocolAddr" -> arpPacket.getHeader.getDstProtocolAddr.toString,
    "hardwareType" -> arpPacket.getHeader.getHardwareType.name,
    "operation" -> arpPacket.getHeader.getOperation.name,
    "protocolType" -> arpPacket.getHeader.getProtocolType.name
  )

  private def getIpPacketMetrics(ipPacket: IpPacket) = Json.obj(
    "srcAddr" -> ipPacket.getHeader.getSrcAddr.getHostAddress,
    "dstAddr" -> ipPacket.getHeader.getDstAddr.getHostAddress,
    "dstIsMc" -> ipPacket.getHeader.getDstAddr.isMulticastAddress,
    "protocol" -> ipPacket.getHeader.getProtocol.name,
    "version" -> ipPacket.getHeader.getVersion.name
  )

  private def getTcpPacketMetrics(tcpPacket: TcpPacket) = Json.obj(
    "srcPort" -> tcpPacket.getHeader.getSrcPort.valueAsInt,
    "srcPortName" -> tcpPacket.getHeader.getSrcPort.name,
    "dstPort" -> tcpPacket.getHeader.getDstPort.valueAsInt,
    "dstPortName" -> tcpPacket.getHeader.getDstPort.name
  )

  private def getUdpPacketMetrics(udpPacket: UdpPacket) = Json.obj(
    "srcPort" -> udpPacket.getHeader.getSrcPort.valueAsInt,
    "srcPortName" -> udpPacket.getHeader.getSrcPort.name,
    "dstPort" -> udpPacket.getHeader.getDstPort.valueAsInt,
    "dstPortName" -> udpPacket.getHeader.getDstPort.name
  )

}
